### PR TITLE
Fix low-rank convergence criterion

### LINKED ISSUE #495: `converged` flag compatibility with `min_iterations` logic
**Describe the bug**
1. Given `min_iterations > inner_iterations` the final costs array will contain `inf`s ([710, 717-720, 731](https://github.com/ott-jax/ott/blob/036add26e4f4391479905e71df910f691f94478c/src/ott/solvers/quadratic/gromov_wasserstein_lr.py#L710-L731)).
2. The `inf`s imply `converged=False` following _all_is_finite_  condition ([201](https://github.com/ott-jax/ott/blob/036add26e4f4391479905e71df910f691f94478c/src/ott/solvers/quadratic/gromov_wasserstein_lr.py#L201)).

**To Reproduce**
```{python}
rngs = jax.random.split(jax.random.PRNGKey(0), 4)
n, m, d1, d2 = 5_000, 2_500, 1, 2
x = jax.random.uniform(rngs[0], (n, d1))
y = jax.random.uniform(rngs[1], (m, d2))
xx = jax.random.uniform(rngs[2], (n, d2))
yy = jax.random.uniform(rngs[3], (m, d2))

geom_x = PointCloud(x)
geom_y = PointCloud(y)
geom_xy = PointCloud(xx, yy)
prob = QuadraticProblem(geom_x, geom_y, geom_xy)
solver = jax.jit(LRGromovWasserstein(
     rank=2, 
     min_iterations=100, 
     inner_iterations=10, 
     max_iterations=2_000,
     progress_fn = utils.default_progress_fn()
))

ot_gwlr = solver(prob)

print(f"converged? {ot_gwlr.converged}")
print(f"costs:{ot_gwlr.costs[np.where(ot_gwlr.costs != -1)[0]]}")
```

**Expected behavior**
Convergence evaluation will ignore `cost` values not computed. 

**Screenshots**
Screenshot 2024-02-27 at 12 15 49

### PR DESCRIPTION
**Pull request by michalk8**
closes #495

### CODE REVIEW

File: src/ott/solvers/linear/sinkhorn_lr.py [hunk 1]
```diff
@@ -687,7 +682,10 @@ def one_iteration(
         lambda: state.reg_ot_cost(ot_prob, epsilon=self.epsilon),
         lambda: jnp.inf
     )
-    error = state.compute_error(previous_state)
+    error = jax.lax.cond(
```
Comment (marcocuturi) [inline src/ott/solvers/linear/sinkhorn_lr.py:685]: good catch!

**Review by marcocuturi — APPROVED** [review]
This LGTM!

I am convinced that we need to get rid of the `(1/ gamma**2)` rescaling in line 60 of `sinkhorn_lr.py`. This rescaling might make sense for a theoretical analysis, but does not make sense for a practical convergence analysis.

tagging @meyerscetbon who might have an opinion, but, for instance, setting large `gamma`, e.g. to 100 vs. 10, makes the same criterion be rescaled "optimistically" (smaller) by 1e2! with a threshold of 1e-3 that we use by default, this makes absolutely no sense, and might explain erratic behavior.

**Comment by marcocuturi** [comment]
I suggest to:
- remove this factor `gamma^2` in the error. With the default `gamma`, this was dividing the criterion by 100, and hence applying a very large `threshold=1e-1`. 
- keep the default `threshold` parameter in `LRSinkhorn`. It's inheriting currently the `1e-3` setting of `Sinkhorn`. When removing `gamma^2` this sounds like a reasonable value (although we might want to make it size dependent, as should be done soon with `Sinkhorn`)

File: src/ott/solvers/linear/sinkhorn_lr.py (hunk 1 above)
Comment (meyerscetbon) [reply src/ott/solvers/linear/sinkhorn_lr.py:685]: The convergence criterion (as currently implemented) only makes “theoretical” sense when the gradient step is constant throughout the iterations. In that case, for each choice of gamma, we could adapt the desired approximation error (by considering gamma^2 \times esp) as suggested Marco. This will be equivalent as the current procedure and we are additionally eliminating an indeterminacy.

However, as we are using a rescaled gradient step, we should, in theory, monitor 1/gamma_k^2 (err_Q + err_R + err_g). And if gamma_k^2 converges towards a constant (close to 1) then it becomes equivalent to only monitor the errors. We could also apply the same idea as in the constant gradient-step case, by monitoring gamma_k^2 \times eps instead.

I hope this helps a little.

**Comment by meyerscetbon** [comment]
The convergence criterion (as currently implemented) only makes “theoretical” sense when the gradient step is constant throughout the iterations. In that case, for each choice of gamma, we could adapt the desired approximation error (by considering gamma^2 \times esp) as suggested Marco. This will be equivalent as the current procedure and we are additionally eliminating an indeterminacy.

However, as we are using a rescaled gradient step, we should, in theory, monitor 1/gamma_k^2 (err_Q + err_R + err_g). And if gamma_k^2 converges towards a constant (close to 1) then it becomes equivalent to only monitor the errors. We could also apply the same idea as in the constant gradient-step case, by monitoring gamma_k^2 \times eps instead.

I hope this helps a little.

**Comment by marcocuturi** [comment]
thanks @meyerscetbon ! I am not sure what you advocate though, practically speaking :) It seems to me that the sum of the 3 KLs is more natural as a quantity to monitor, independently of the stepsize that's chosen. Is there any reason why, intuitively, the termination criterion should be scaled with stepsize? For instance, if the first `gamma=1000` (small stepsize) we will immediately converge in one iteration in principle. This seems counterintuitive to me.

**Comment by meyerscetbon** [comment]
I was trying to say that in the proof, we look at the convergence of (1/\gamma_k^2) * err (where err are the KL terms) and gamma_k is not assumed to be constant (indeed we only assume that it is constant at the end of the proof in order to obtain a sufficient condition of convergence). Therefore there might be some cases where the convergence of (1/\gamma_k^2) * err does not imply the convergence of err. However I think we can safely assume that gamma_k will converge towards a constant and therefore we can only monitor the error term as you suggested.

I was also trying to say that when gamma is constant along the iterations, then rescaling the error by  (1/\gamma^2) or not is the same in term of convergence, and we can, as you suggested, remove (1/gamma^2) from the criterion and only monitor the error term. 

Sorry for the confusion.

**Comment by giovp** [comment]
thanks for the discussion all, very interesting. I've just tested on some of the failing tests we have in moscot and now the convergence seems to be set correctly (returning True when it should be).

**Comment by selmanozleyen** [comment]
These are the related issues:
- https://github.com/theislab/moscot/issues/703
- https://github.com/theislab/moscot/issues/678
